site-content/source/modules/ROOT/pages/development/testing.adoc

```diff
@@ -1,21 +1,73 @@
 :page-layout: basic
 
-== Testing
+= Testing
 
 Creating tests is one of the most important and also most difficult
 parts of developing Cassandra. There are different ways to test your
 code depending on what you're working on.
 
-=== Unit Testing
+Cassandra tests can be divided into three main categories, based on the
+way how they are executed:
 
-The simplest test to write for Cassandra code is a unit test.
-Cassandra uses JUnit as a testing framework and test cases
-can be found in the `test/unit` directory.
-Ideally, you’d be able to create a unit test for your implementation that would exclusively cover the class you created (the unit under test).
-Unfortunately, this is not always possible, because Cassandra doesn’t have a very mock friendly code base.
-Often you’ll find yourself in a situation where you have to use an embedded Cassandra instance to interact with your test.
-If you want to make use of CQL in your test, you can extend CQLTester and use
-some of the convenient helper methods, as shown here:
+- *JVM tests* - tests implemented in Java and being a part of the
+Cassandra project. JVM tests can be divided internally into three
+subcategories - regular tests, distributed tests, and micro-benchmarks.
+Technically both regular and distributed tests are just Java JUnit tests
+but have different requirements and test Cassandra differently. In
+Cassandra 4.0, there are ~7k JVM test cases
+
+- *CQLSH tests* - CQLSH tests are Python tests written with the Nose
+test framework. They simply verify the CQLSH client that can be found in
+the bin directory. They aim at verifying CQLSH specific behavior like
+output formatting, autocompletion, parsing, etc.). Those tests assume
+there is a Cassandra cluster that is already running. Each test case
+starts the CQLSH client as a subprocess, issues some commands, and
+verifies the outcome returned by CQLSH to the console. In Cassandra 4.0,
+there are ~90 test cases for this category
+
+- *Python Distributed tests (dtests)* - Python distributed tests are
+implemented on top of the PyTest framework and located outside the main
+Cassandra project in the separate repository
+https://github.com/apache/cassandra-dtest[apache/cassandra-dtest]. They
+are testing Cassandra by starting a cluster consisting of one or more
+nodes and performing various operations against that cluster. After the
+test procedure, they verify the operation results, logs (no unexpected
+errors), and the ultimate cluster state. Because they utilize the
+https://github.com/riptano/ccm[CCM library] internally, unlike CQLSH
+tests, no running cluster is assumed. Python Distributed tests are
+Cassandra version agnostic, so you run the same set of tests against
+whichever Cassandra version you want. The tests have various assumptions
+internally, making the framework skip the tests that are not relevant
+to the Cassandra version you are testing
```

Comment:
I guess… there's two different perspectives here and they are both valuable feedback to the newcomer. A) how the different tests are run (and categorising them that way…), and B) how the project community classifies them (more type of test and its application, rather than just the approach of running them…)

Reply:
Well, indeed, that page about testing is focused about how to run tests. Tests which are run in the same way are grouped together because they share many details. Description of the community classification would be great addition, perhaps in a separate section or somewhere in Getting Started page